### PR TITLE
Dataset api

### PR DESCRIPTION
**Pull request by gdj0nes**

### CODE REVIEW

File: muffnn/core.py [hunk 1]
```diff
@@ -328,6 +328,7 @@ def _build_tf_graph(self):
         """
         self._set_up_graph()
         self._saver = tf.train.Saver()
+        # TODO: presonally thing session calls should be within with clauses
```
Comment (beckermr) [inline muffnn/core.py]: As discussed, we are leaving this as is.

File: muffnn/mlp/base.py [hunk 2]
```diff
@@ -138,6 +138,7 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
             # Instantiate the graph.  TensorFlow seems easier to use by just
             # adding to the default graph, and as_default lets you temporarily
             # set a graph to be treated as the default graph.
+            # TODO: why is graph_.as_default being called, isn't there just one graph?
```
Comment (beckermr) [inline muffnn/mlp/base.py]: Again leaving this as is.

File: muffnn/mlp/base.py [hunk 3]
```diff
@@ -156,30 +157,32 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
+                    # TODO: figure out how to catch end of epoch to keep this
+                    if monitor:
+                        # TODO what is monitor
+                        stop_early = monitor(epoch, self, {'loss': obj_val})
+                        if stop_early:
+                            _LOGGER.info(
+                                "stopping early due to monitor function.")
+                            return self
+                except tf.errors.OutOfRangeError:
+                    break
```
Comment (beckermr) [inline muffnn/mlp/base.py]: We want to preserve more of this loop so the logging works similarly.

File: muffnn/mlp/base.py [hunk 4]
```diff
@@ -156,30 +157,32 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
+                              feed_dict={self._features_placeholder: X,
+                                         self._labels_placeholder: y})
+
+            while True:
+                try:
+                    # TODO: add a global step variable
+                    # Taining step
+                    obj_val, _ = self._session.run([self._obj_func, self._train_step])
+
+                    # TODO: reduce the number of saved values, this should be replaced with a summary writer
```
Comment (beckermr) [inline muffnn/mlp/base.py]: IDK what this means. None of the values are saved. They are output from the session into python objects.

File: muffnn/mlp/base.py [hunk 5]
```diff
@@ -156,30 +157,32 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
+                    _LOGGER.debug("objective: %.4f, epoch: %d, idx: %d", obj_val)
+
+                    # TODO: figure out how to catch end of epoch to keep this
+                    if monitor:
+                        # TODO what is monitor
```
Comment (beckermr) [inline muffnn/mlp/base.py]: See the doc strings on the MLPClassifier methods. These are functions that allow for early stopping and custom monitoring in the sklearn API.

File: muffnn/mlp/base.py [hunk 6]
```diff
@@ -156,30 +157,32 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
+            while True:
+                try:
+                    # TODO: add a global step variable
+                    # Taining step
+                    obj_val, _ = self._session.run([self._obj_func, self._train_step])
+
+                    # TODO: reduce the number of saved values, this should be replaced with a summary writer
+                    _LOGGER.debug("objective: %.4f, epoch: %d, idx: %d", obj_val)
+
+                    # TODO: figure out how to catch end of epoch to keep this
```
Comment (beckermr) [inline muffnn/mlp/base.py]: The number of training steps per epoch can be precomputed. If this does not work, we can just run one epoch at a time even when the user asks for more than one. This will still eliminate overhead but would allow for us to catch the end of each one.

File: muffnn/mlp/base.py [hunk 7]
```diff
@@ -247,14 +250,39 @@ def _set_up_graph(self):
...
+
+            # TODO: pray the following is not what we have to do
+            (x, y, z), self._input_targets = iterator.get_next()
+
             # t will be the current layer as we build up the graph below.
-            t = tf.SparseTensor(self._input_indices, self._input_values,
-                                self._input_shape)
+            t = tf.SparseTensor(x, y, x)
+            # TODO: if map works
+            self._input_values, self._input_targets = iterator.get_next()
```
Comment (beckermr) [inline muffnn/mlp/base.py]: Per our discussion I think we just use the simple version of this for now and can change it later if needed.

File: muffnn/mlp/base.py [hunk 8]
```diff
@@ -286,31 +314,6 @@ def _set_up_graph(self):
...
-
-        if y is None:
-            # If y is None, then we are doing prediction and should fix
-            # dropout.
-            feed_dict[self._keep_prob] = 1.0
-        else:
-            feed_dict[self.input_targets_] = y
-            feed_dict[self._keep_prob] = self.keep_prob
-
-        return feed_dict
```
Comment (beckermr) [inline muffnn/mlp/base.py]: We need to keep this function in order to control the behavior for dropout and predictions. We can adopt it to build the proper feed dictionaries for various cases.

File: muffnn/mlp/mlp_classifier.py [hunk 9]
```diff
@@ -111,6 +111,9 @@ def _init_model_output(self, t):
                 t = tf.nn.dropout(t, keep_prob=self._keep_prob)
             t = affine(t, output_size, scope='output_layer')
 
+        # TODO: discuss where to move the instantiaion of self.input_targets_
+        # TODO: it need to be moved or the init_model_output needs to take an addition input
+
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: To preserve the API of the internal objects, we need to leave this here. IDK if it is a problem but I think it should be ok.

**Comment by beckermr** [comment]
A few other outstanding issues:

1. Do we want to use the Datasets API with the compute output function too?
2. Do we need to make changes in the MLPRegressor?
3. What about the rest of the classes like the autoencoder and the FM?

File: muffnn/mlp/base.py (hunk 7 above)
Comment (gdj0nes) [reply muffnn/mlp/base.py]: The only issue is that we need to keep the dataset function chaining inside the `is_sparse` block so that we can set `t` correctly

File: muffnn/mlp/mlp_classifier.py (hunk 9 above)
Comment (gdj0nes) [reply muffnn/mlp/mlp_classifier.py]: I'll try to think of a work around, but `self.input_targets_` needs to be passed into the dataset API. I'm thinking that maybe we can zip up the datasets after the fact. However I'm not exactly sure how the synchronization would work in such a case.

File: muffnn/mlp/base.py (hunk 7 above)
Comment (beckermr) [reply muffnn/mlp/base.py]: right yep.

File: muffnn/mlp/mlp_classifier.py (hunk 9 above)
Comment (beckermr) [reply muffnn/mlp/mlp_classifier.py]: hmmmmm. I need to see the code more to understand this. A close to working version with just this issue would help me.

**Comment by beckermr** [comment]
Whoa. This is a blast from the past. Hope all is well!
